### PR TITLE
Translation Table

### PR DESCRIPTION
**Pull request by milenejo**
Code to use translation_table argument in bricks2marble. Can be used by adding argument --translation_table [int]. The default is None, where originally pre-defined starts and stops will be used. When setting translation_table, these pre-defined starts and stops will be overwritten.

### CODE REVIEW

**Comment by MarioStanke** [comment]
It looks like you have added the `translation_table` option to annotate but not to `vipsania train`. Is that so? 
The main use case for species nonstandard genetic code is to train vipsania on the very genome, though. Can you check and add this if not present?

**Comment by milenejo** [comment]
The translation_table option has now been added to train as well. It functions the same as for annotate, and can be set using --translation_table [int]

**Comment by MarioStanke** [comment]
Great, I have tested it and confirmed the expected behaviour on _Tetrahymena thermophila_ (translation_table 6) and _Drosophila melanogaster_ (`translation_table` 1) using single-genome training and whole genome evaluation against a referene annotation. 
The new code with  `--translation_table 6` produces 9.67% glumine amino acids (Q), almost exactly as in the RefSeq annotation (9.62%). The old code was way off in terms of Q content and accuracy.
There is no regression in the standard genetic code case (D. mel.), both versions work similarly fine.

**Review by MarioStanke — CHANGES_REQUESTED** [review]
Assume, a model has been trained with `translation_table` 6 for ciliates. One then runs
`vipsania annotate <tt6-model> ciliate.fa -o out.gff3`
This predicts the genes correctly, but then silently deletes them afterwards as containing in-frame stops. The false translation table would also be used to generate a protein.fa. The users would likely conclude that Vipsania does not work.
The translation_table should be taken from the config of the hmm if not specified on the command line.

Please also adjust docs/training.md to the command line flag. Currently, it instructs users to edit a json file. It should reflect the way you tested this new function. As  Richard and I have tested the accuracy, it is sufficient if you just do a very short training to confirm in the files that there are no warnings or errors and the right tt and set of stop codons is used for prediction, filtering and translation, when it is used as you specify in the training.md.

**Comment by milenejo** [comment]
The translation_table info can now also be taken from the config. I have edited docs/training.md accordingly.

**Review by MarioStanke — APPROVED** [review]
All right. I added some documentation and also bumped the version.

**Comment by guyleonard** [comment]
Tried this branch on *Paramecium bursaria* 186b (GCA_053455385.1, table 6) as a
second test case alongside *Tetrahymena*, using the `translation_table` branches
of both Vipsania and bricks2marble. The b2m helpers return the expected values:

```python
>>> from bricks2marble.struct.start_stop_codons import get_start_codons, get_stop_codons
>>> get_stop_codons(6), get_start_codons(6)
([('TGA', 1)], [('ATG', 1)])
```

I was not able to get as far as evaluating a table-6 model, because finetuning on
this genome produces an empty annotation whether or not `--translation_table` is
given - filed separately as #5 , and not a problem with this PR.

`--translation_table 6` is carried correctly into the model config by
`annotate --finetune`:

```python
>>> json.load(open("finetuning_1/config.json"))["model"]["hmm"]
{'translation_table': 6, 'start_codons': [['ATG', 1.0]], 'stop_codons': [['TGA', 1.0]], ...}
```

Annotating with the stock `Alveolata` weights and `--translation_table 6` runs
normally and predicts genes (39 on a test scaffold). That run does not really
exercise the table though: the released model's own config specifies the three
standard stop codons

```python
'stop_codons': [['TAG', 0.34], ['TAA', 0.33], ['TGA', 0.33]]
```

and its predicted CDS contain no in-frame TAA/TAG at all (0 in 2,636 codons),
where the reference annotation for this genome has 6.90 per 100 codons.

I've added a couple of issues related to this. Really cool program otherwise and have been testing it on some less weird protists and they have been coming out really well! :)

**Comment by milenejo** [comment]
Currently, using a model that specifies a different translation table/different start and stop codons than the given `--translation_table` argument is not supported. With the most recent commit `5bb3744`, an error should be raised when attempting to annotate with `Alveolata` weights and `--translation_table 6`.

**Comment by guyleonard** [comment]
Thanks, I re-tested it and it now correctly gives a warning!

I have been able to actually rerun and trained a table 6 model completely from scratch (vipsania train configs/base_25M.json --translation_table 6, no fork) on Paramecium bursaria plus 18 other public Paramecium genomes (assemblies from ParameciumDB: P. biaurelia, caudatum, decaurelia, dodecaurelia, jenningsi, novaurelia, octaurelia, pentaurelia, primaurelia, quadecaurelia, sonneborni, tetraurelia, sexaurelia, plus 5 other bursaria strains), ~1.25 GB total on an NVIDIA A40.

I tried a deliberately small run (20 epochs × 400 steps, final accuracy 0.428) and I get some usable output against my Pabu reference:
 
|                                  | genes  | gene-locus Sn / Pr | CDS-bp Sn / Pr |
|----------------------------------|--------|--------------------|----------------|
| multi-species, 20 ep             | 18,099 | 0.765 / 0.666      | 0.674 / 0.694  |
| Paramecium bursaria alone, 10 ep | 33,198 | 0.866 / 0.594      | 0.808 / 0.583  |

The multi-species set clearly helps calibration (higher precision, closer to the true gene count of 21,195) versus training on Pabu alone, which I guess over-predicts. Super! I am gonna do some more testing. Thanks for your response and cool program!

**Comment by irkri** [comment]
Hi @guyleonard, the results look very good to me. Thanks for testing this!
I was about to suggest training a Vispania model from scratch for your issue #5, but your recent results show that you already tried and it works, great.

Annotating species that require a different translation table with one of the shipped pretrained model was never planned/tested, so training from scratch is the way to go. The 2M or 10M variants should produce similar results on your genomes, as they are very small. Finetuning gives best improvements on larger species, so I expect this to not have a large impact on annotation quality. Nevertheless, you could try annotating Paramecium bursaria with finetuning using your custom multi-species model, too. 

I will soon push a fix on bricks2marble for #6 and probably adapt the log to show how many genes were removed by postprocessing for #5.
